src/server/mod.rs

```diff
@@ -368,24 +409,25 @@ impl<O: Output> LsService<O> {
                 notifications::DidChangeConfiguration,
                 notifications::DidChangeWatchedFiles,
                 notifications::Cancel;
-            requests:
+            blocking_requests:
                 ShutdownRequest,
                 InitializeRequest,
-                requests::Definition,
-                requests::References,
-                requests::Completion,
                 requests::ResolveCompletion,
-                requests::Rename,
-                requests::DocumentHighlight,
                 requests::ExecuteCommand,
+                requests::Deglob,
+                requests::Formatting,
+                requests::RangeFormatting;
+            requests:
+                requests::Rename,
                 requests::CodeAction,
+                requests::DocumentHighlight,
                 requests::FindImpls,
-                requests::Deglob,
                 requests::Symbols,
+                requests::Hover,
                 requests::WorkspaceSymbol,
-                requests::Formatting,
-                requests::RangeFormatting,
-                requests::Hover;
+                requests::Definition,
+                requests::References,
+                requests::Completion;
```

Comment:
What's the motivation for having blocking and non-blocking requests?

Reply:
I think all `requests::*` should be stdin-non-blocking. `ShutdownRequest` and `InitializeRequest` however benefit. The former requires mutable access to the server state, and the latter is meant to be handled before any other requests.

In general though we should avoid implementing new requests as blocking.

Reply:
The remaining blocking requests send further requests/notification so currently require access to the `Output`. This needs to be reworked a little to make them non-blocking. Non-blocking requests have no access to the `Output` and have to respond what they want to do to the dispatch-worker.

Currently requests can respond with a result or error. This can be extended to support additional requests to the client. But I thought I'd leave this until later and see how you liked the PR as is.